### PR TITLE
Survival probability: Intermittency, Step, Residues, Overlapping SP Selection Example

### PR DESCRIPTION
**Pull request by bieniekmateusz**
Fixes #
 - "Step" parameter does not load frames which are not used (dependant on the value of tau_max). This is great of sampling long trajectories. 

Changes made in this Pull Request:
 - residue-wise SP can be carried out
 - Intermittency is defined as a number of consecutive absences allowed 
 - Intermittency combined with Step lead to consistent behaviour
 - the raw data are made available (helps with testing, allows building upon the values)
 - simplified and extended documentation to include SP around overlapping areas (for example, headgroups in lipids)
 - Test Cases for all new functionalities have been added 

PR Checklist
------------
 - [X] Tests?
 - [X] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

We generalised the code to be applicable to other situations which is why we would like to ask if we should move it out of waterdynamics.py to its own file (survivalprobability.py in analysis). 

The intermittency, for the sake of simplicity, is defined in terms of consecutive absences. This is calculated in a single pass over the data. 

We've used the new functionalities and plan to publish on the topic to discuss the usefulness of these small changes. 

This request is done on behalf of me and @p-j-smith

### CODE REVIEW

**Comment by bieniekmateusz** [comment]
Hi there, I only modified two files, however, I worked on the "develop" branch by mistake. So I tried to rebase the code to bring my commits to the front, which is why the review picks up so many files. 

I'll see if I can move the commits differently to avoid what happened here.

**Comment by bieniekmateusz** [comment]
Hi again, I see that there are some functionalities which could help with organising the commits: git cherry-pick in particular. Let me know if it is necessary and then I'll further look into it.

**Comment by orbeckst** [comment]
Looks like a clean diff, so no need to play more with `git rebase` and friends.

**Comment by orbeckst** [comment]
I merged latest develop into your branch so that the docs build correctly and don't fail your build.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 1]
```diff
@@ -297,13 +297,50 @@
...
+  sp_timeseries.insert(1, 0)
+
   # plot
   plt.xlabel('Time')
   plt.ylabel('SP')
   plt.title('Survival Probability')
   plt.plot(tau_timeseries, sp_timeseries)
   plt.show()
 
+Another example applies to the situation where you work with many different "residues".
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Nice example. Good to have it here for now, eventually we would probably move it into a section with examples #1175 .

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 2]
```diff
@@ -297,13 +297,50 @@
...
+Here we calculate the SP of a potassium ion around each lipid in a membrane and
+average the results. In this example, if the SP analysis were run without treating each lipid
+separately, potassium ions may hop from one lipid to another and still be counted as remaining
+in the specified region. That is, the survival probability of the potassium ion around the
+entire membrane will be calculated.
+
+Note, for this example, it is advisable to use .Universe(in_memory=True) to ensure that the
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: use backticks for formatting
```reST
to use `Universe(in_memory=True)` to ensure
```
and don't use a leading do. It's ok to just name the class.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 3]
```diff
@@ -297,13 +297,50 @@
...
+simulation does not loaded into RAM memory once per lipid::
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Do you mean "... to ensure that the simulations is not loaded into RAM memory once per lipid"?

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 4]
```diff
@@ -297,13 +297,50 @@
...
+
+  import MDAnalysis
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: We customarily use
```python
import MDAnalysis as mda
```
(or at least that's what we should be doing consistently)

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 5]
```diff
@@ -1231,22 +1266,78 @@ def print(self, verbose, *args):
         elif verbose:
             print(args)
 
-    def run(self, tau_max=20, start=0, stop=None, step=1, verbose=False):
+    def _correct_intermittency(self, intermittency, selected_ids):
         """
-        Computes and returns the survival probability timeseries
+        Pre-process Consecutive Intermittency
+        :param intermittency: the max gap allowed and to be corrected
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Use numpy/Napoleon formatting, see https://github.com/MDAnalysis/mdanalysis/wiki/WritingDocumentation#numpy-style

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 6]
```diff
@@ -1231,22 +1266,78 @@ def print(self, verbose, *args):
...
+        :param selected_ids: modifies the selecteded IDs in place by adding atoms which left for <= :param intermittency
+        """
+        self.print('Correcting the selected IDs for intermittancy (gaps). ')
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Was `self.print()` already in the class?

(We should be consistently using logging... but I think that's another PR for this module.)

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 7]
```diff
@@ -1231,22 +1266,78 @@ def print(self, verbose, *args):
         elif verbose:
             print(args)
 
-    def run(self, tau_max=20, start=0, stop=None, step=1, verbose=False):
+    def _correct_intermittency(self, intermittency, selected_ids):
         """
-        Computes and returns the survival probability timeseries
+        Pre-process Consecutive Intermittency
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Please explain what this function does. I have no idea what "intermittency" means here.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 8]
```diff
@@ -1231,22 +1266,78 @@ def print(self, verbose, *args):
...
+        :param intermittency: the max gap allowed and to be corrected
+        :param selected_ids: modifies the selecteded IDs in place by adding atoms which left for <= :param intermittency
+        """
+        self.print('Correcting the selected IDs for intermittancy (gaps). ')
+        if intermittency == 0:
+            return
+
+        # If the atom is absent for a number of frames equal or smaller
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Stuff like this should be in actual docs.

Comment (orbeckst) [reply package/MDAnalysis/analysis/waterdynamics.py]: Or more precisely: the comments here are very good, but the docstring should tell a user what the function does (what problem does it solve). The comments should explain the how – the ideas behind the algorithm.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 9]
```diff
@@ -1231,22 +1266,78 @@ def print(self, verbose, *args):
...
             Zero-based index of the first frame to be analysed
-        stop : int
+        stop : int, optional
             Zero-based index of the last frame to be analysed (inclusive)
-        step : int
-            Jump every `step`'th frame
-        tau_max : int
+        step : int, optional
+            Jump every `step`'th frame. This is compatible but independant of the taus used, and it is good to consider
+            using the  step as tau_max to remove the overlap. Note that step and taus are consistent with intermittency.
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: put all references parameters in single backticks such as
```reST
Jump every `step` -th frame. This is compatible but independant of the taus used, and it is good to consider
using the  `step` as `tau_max` to remove the overlap.
```
Note that you need a space after the backticks.

Build the docs locally and check that your docs look ok, see https://github.com/MDAnalysis/mdanalysis/wiki/WritingDocumentation#generating-the-docs

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 10]
```diff
@@ -1231,22 +1266,78 @@ def print(self, verbose, *args):
...
+        tau_max : int, optional
             Survival probability is calculated for the range :math:`1 <= \tau <= tau_max`
-        verbose : Boolean
-            Overwrite the constructor's verbosity
+        residues : Boolean, optional
+            If true, the analysis will be carried out on the residues (.resids) rather than on atom (.ids).
+            A single atom is sufficient to classify the residue as within the distance.
+        intermittency : int, optional
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Good explanation.

**Comment by orbeckst** [comment]
> We generalised the code to be applicable to other situations which is why we would like to ask if we should move it out of waterdynamics.py to its own file (survivalprobability.py in analysis).

I like the idea of moving it into its own module. How about doing this as a separate PR once this one is merged? The smaller a PR the better.

**Comment by orbeckst** [comment]
I broke travis #2233 so once PR #2234 is merged we need to again merge develop into this branch, sorry.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 2 above)
Comment (bieniekmateusz) [reply package/MDAnalysis/analysis/waterdynamics.py]: Corrected.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 1 above)
Comment (bieniekmateusz) [reply package/MDAnalysis/analysis/waterdynamics.py]: Noted. I'll have a read later.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 6 above)
Comment (bieniekmateusz) [reply package/MDAnalysis/analysis/waterdynamics.py]: Yes, we've created self.print() in the previous PR. I'll look into logging for the next PR. Thanks

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 8 above)
Comment (bieniekmateusz) [reply package/MDAnalysis/analysis/waterdynamics.py]: Good point. I've moved the actual documentation to the top and introduced minor corrections in the comments.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 7 above)
Comment (bieniekmateusz) [reply package/MDAnalysis/analysis/waterdynamics.py]: Addressed: moved the comment to the documentation since it explains it well.

**Comment by bieniekmateusz** [comment]
The requested changes have been introduced. I've built the documentation and it appears fine. Cheers

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 11]
```diff
@@ -1273,15 +1273,42 @@ def run(self, tau_max=20, start=0, stop=None, step=1, verbose=False):
...
-            selected_ids.append(set(self.universe.select_atoms(self.selection).ids))
+        if step <= tau_max + 1:
+            # load all frames
+            for ts in self.universe.trajectory[start:stop]:
+                self.print(verbose, "Loading frame:", ts)
+                selected_ids.append(set(self.universe.select_atoms(self.selection).ids))
+        else:
+            # skip frames that will not be used
+            # Example: step 5 and tau 2: L, L, L, S, S, L, L, L, S, S, ... with L = Load, and S = Skip
+            loaded_counter = 0
```
Comment (alejob) [inline package/MDAnalysis/analysis/waterdynamics.py]: Why would you like to do this? Maybe I'm missing something, but when you calculate the Survival Probability, the definition says that you should consider only the particles that REMAIN in the selection between time t and t+tau, ie, you should load all the frames between t and t+tau and count in t, then count in t+1, count in t+2..., count in t+tau, counting only the particles that were in the selection in the previous step, so I don't understand why would you like to skip some frames when step is larger than the tau_max + 1. Maybe an other example could help to clarify this.

**Comment by alejob** [comment]
> > We generalised the code to be applicable to other situations which is why we would like to ask if we should move it out of waterdynamics.py to its own file (survivalprobability.py in analysis).
> 
> I like the idea of moving it into its own module. How about doing this as a separate PR once this one is merged? The smaller a PR the better.

It sounds good, but how could we avoid to broke the old code? There is people that has already implemented scripts with waterdynamics.survivalprobability(), so the idea is to avoid broke that code. Maybe some kind of link inside the MDA code.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 11 above)
Comment (bieniekmateusz) [reply package/MDAnalysis/analysis/waterdynamics.py]: Thanks @alejob 

It is always the case that we load all the frames between t and t+tau. The step foremost allows to remove the overlap of the measurement. I.e. a step larger than tau_max may be required to ensure that subsequent t0s are not correlated. To rephrase, if an atom survives from t to t+tau, then for the next 20 frames, the data for tau=1 will have 20 data points where the atoms survives.

Then there is a performance regarding sampling. If someone would like to scan through a long simulation, skipping frames is a huge performance gain. This data can be also used for discussion for equilibration and stability of the system.

Mat and @p-j-smith

**Comment by bieniekmateusz** [comment]
So moving it to another file would be a new PR in the future. 

We could say that the code is moved and simply point the user to import the new path - if someone tries to use it. Luckily the function itself and how it works has not changed so it should be a minor inconvinience for the users, hopefully.

**Comment by orbeckst** [comment]
Breaking user code is not something that we do lightly. 

In this case it will be easy: we just import the function into water dynamics and everything else will just be able to use it as before. The call signature did not change (?) and all the old tests for the function still pass, don't they?

In any case, that's all stuff we can discuss in a new issue/PR. 

--
Oliver Beckstein
email: orbeckst@gmail.com

**Comment by orbeckst** [comment]
Merged develop to get the travis fix from PR #2234

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 5 above)
Comment (bieniekmateusz) [reply package/MDAnalysis/analysis/waterdynamics.py]: Corrected.

**Comment by bieniekmateusz** [comment]
I think I've corrected everything, please let me know if I missed anything.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 11 above)
Comment (alejob) [reply package/MDAnalysis/analysis/waterdynamics.py]: Great, now I understand, and this look pretty good. Thanks for your improvement.

**Comment by bieniekmateusz** [comment]
To give you heads up, after this is merged we will shortly send you another PR with a common autocorrelation function between the waterdynamics.HydrogenBondLifetimes and waterdynamics.SurvivalProbability. 

@richardjgowers noted that it would be good to formalise and extract the common autocorrelation code (PR #1995). After careful considerations with @p-j-smith we now have a separate autocorrelation function that has intermittency built into it, and being independent, is easily testable. The test cases will be well-defined in a similar manner to the ones we've used for the SurvivalProbability. We've already verified that waterdynamics.HydrogenBondLifetimes behaves exactly the same way our autocorrelation does (continuous). These changes should to a big extent simplify the module. In the case of HydrogenBondLifetimes it would mean a simple function call replacing the entire class. This will help simplifying the documentation too (PR #2181)

Work done with @p-j-smith.

File: package/CHANGELOG [hunk 12]
```diff
@@ -15,7 +15,7 @@ The rules for this file:
 ------------------------------------------------------------------------------
 mm/dd/yy micaela-matta, xiki-tempula, zemanj, mattwthompson, orbeckst, aliehlen,
          dpadula85, jbarnoud, manuel.nuno.melo, richardjgowers, mattwthompson,
-         ayushsuhane, picocentauri, NinadBhat
+         ayushsuhane, picocentauri, NinadBhat, bieniekmateusz, p-j-smith
```
Comment (orbeckst) [inline package/CHANGELOG]: Are @bieniekmateusz and @p-j-smith in AUTHORS?

**Comment by orbeckst** [comment]
@alejob if you're happy with the PR could you please also add a formal "approve" review? Thanks.

File: package/CHANGELOG (hunk 12 above)
Comment (bieniekmateusz) [reply package/CHANGELOG]: We, both of us. We're doing pair-programming.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 13]
```diff
@@ -1231,27 +1265,89 @@ def print(self, verbose, *args):
...
+            If true, the analysis will be carried out on the residues (.resids) rather than on atom (.ids).
+            A single atom is sufficient to classify the residue as within the distance.
+        intermittency : int, optional
+            The maximum number of consecutive frames for which an atom can leave but be counted as present if it returns
+            at the next frame. An intermittency of `0` is equivalent to a continuous survival probability, which does
+            not allow for the leaving and returning of atoms. For example, for `intermittency=2`, any given atom may
+            leave a region of interest for up to two consecutive frames yet be treated as being present at all frames.
+            The default is continuous (0).
+        verbose : Boolean, optional
+            Print the progress to the console
```
Comment (alejob) [inline package/MDAnalysis/analysis/waterdynamics.py]: I like this, it gives great flexibility to SP.

**Comment by orbeckst** [comment]
Thanks @alejob !

I resolved the conflict in the CHANGELOG and we'll just let CI confirm that everything is still good.

**Comment by orbeckst** [comment]
Congratulations @bieniekmateusz and @p-j-smith , merged! Thanks for your hard work (and patience).

Thanks @alejob for reviewing.

**Pull request merged**